### PR TITLE
Feature/TPA

### PR DESCRIPTION
**Pull request by anton-matosov**
Re-implement Throttle PID Attenuation to support per-component setting of attenuation break and slope rate.
Used a bit different function that has preductable curve based on inputs.
You can compare both functions here https://www.desmos.com/calculator/gn4mfoddje

(PR includes new ZMR250 frame rc and mix files)

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Awesome! Do you have some feedback on MC_RATE_I_MODE?

**Comment by anton-matosov** [comment]
I have removed it from this PR, I am still trying to figure out how this code affects flight characteristics

**Comment by kd0aij** [comment]
@anton-matosov Interesting; have you done any testing to show the motivation for and effectiveness of also scaling the I and D gains? Also, what is more predictable about the scaling now?

**Comment by anton-matosov** [comment]
My drone had D oscilations at 0.8+ throttle that I couldn't compensate with existing TPA. That was the main motivation for these changes.
While implementing it I have found that it's hard to predict what will be your PID rate attenuation with max throttle for particular break and slope values. 
With my function you always have (slope at full throttle==PID scale factor) and linear function from breakpoint to full for the scale.
You can see graphs for both functions here https://www.desmos.com/calculator/gn4mfoddje and play with b (break) and r (rate/slope) sliders

**Comment by kd0aij** [comment]
@anton-matosov Thanks,that makes sense. I hadn't noticed a need for decreasing D when tuning and my procedure was just to increase the slope till the oscillations were inaudible. Hence I didn't really care what the actual attenuation at max throttle was, though I did calculate it afterward. With the new scaling, both r and b affect the slope, so it might be best to rename the r parameter to something which more accurately names its function.

**Comment by anton-matosov** [comment]
Agreed. In other flight stacks it's called TPA Rate
I will rename it tonight

**Comment by LorenzMeier** [comment]
@anton-matosov This looks awesome. Is it feasible to enable a safe default for all platforms? Do you have a chance to test with a 450 sized drone?

**Comment by anton-matosov** [comment]
Unfortunately I have only 250 racer and 550 hexa (which weight 2 kg). I was ok to test full throttle with racer, but not comfortable at all to do this with hexa.
IMO default for TPA - is no TPA, as usually PIDs are not tuned to their extremes so there are no oscillations at higher/full throttle.

**Comment by anton-matosov** [comment]
Can anyone please take a look at this PR?

**Comment by anton-matosov** [comment]
Sure, will change to 4080

**Comment by LorenzMeier** [comment]
@anton-matosov Can you rebase on master? Should also fix the flash issue.

**Comment by anton-matosov** [comment]
@LorenzMeier build fails with error

```
/opt/gcc/bin/../lib/gcc/arm-none-eabi/4.9.3/../../../../arm-none-eabi/bin/ld: src/firmware/nuttx/firmware_nuttx section `.data' will not fit in region `flash'
/opt/gcc/bin/../lib/gcc/arm-none-eabi/4.9.3/../../../../arm-none-eabi/bin/ld: region `flash' overflowed by 1172 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
Makefile:166: recipe for target 'px4fmu-v2_default' failed
make[2]: *** [px4fmu-v2_default] Error 1
Makefile:246: recipe for target 'check_px4fmu-v2_default' failed
make[1]: *** [check_px4fmu-v2_default] Error 2
Makefile:246: recipe for target 'check_qgc_firmware' failed
make: *** [check_qgc_firmware] Error 2
```

How can I troubleshoot this? Do others get something similar?

**Comment by LorenzMeier** [comment]
@anton-matosov If you rebase the branch the flash overflow will go away.

Is it really necessary to add the 4th 250 racer config? I'm a bit worried that we have a lot of stale configs and these racers tend to be very close to each other. Is there a way to reduce them to 1-2 configs?

**Comment by kd0aij** [comment]
The roll/pitch P gains for the new ZMR250 config are only about 1/4 of those for the QAV250 and generic250... not very close. Does it hold attitude well in acro mode?

**Comment by LorenzMeier** [comment]
Sorry, I should have been more accurate: I would be surprised if the gains differ dramatically for real frames. If they do I rather suspect insufficient tuning than actual differences.

**Comment by anton-matosov** [comment]
@kd0aij P gains don't even participate in the acro mode, only rates. The main influencer of the attitude hold for acro mode is I rate, I have found it should be in the range of 0.19-0.25 for my frame/motors/esc/battery/props setup.

@LorenzMeier parameters are actually very unique for frame/motors/esc/battery/props combination. If I simply change props I already need to re-tweak parameters (deviation depends on how big is the change).
ZMR250 and QAV250 frames are very similar, but other components are not.
In addition to the frame size we should consider combinations of other components:
 - battery 3s vs 4s (totally different P and D rates) + vendor (different discharge rate and curve which affects TPA settings)
 - esc (different refresh frequency and motor timings and firmware, can drastically affect final thrust and require different P and D rates)
 - motor (generally only KV matters, totally different P and D rates for 2050kv and 2400kv)
 - props (diameter and pitch affect tuning significantly, e.g. 5030 2 blade vs 5051 3 blade requires different P and D rates)
 - final weight also matter

Setup for which zmr250 config was created is as follows:
 - battery: 2200mah 3s
 - motors: 1808 V-Spec 2400kv
 - esc: Afro 20A Race Spec ESC running SimonK
 - props: Gemfan 5030 2 blade

I will put this info into config as well.

I have switched to 3 blades recently and tried 2 different kinds:
 - For DAL 5045 quad become a bit sluggish, so I had to lower D a little bit
 - For RaceKraft RK5051 I also got oscillations on higher throttle and sharp turns, so I will have to adjust P rate and TPA settings (didn't have time to find perfect settings yet).

I have some flight videos captured for current settings and RK5051 props. Once I get a change to process them I will share them in the px4 discussion forums.

**Comment by kd0aij** [comment]
@anton-matosov Of course I was referring to the *rate* P gains; sorry for the ambiguity in my comment.
The roll/pitchrate_p gains for the ZMR250 in this PR are .05, whereas those for the generic 250 are .19

**Comment by anton-matosov** [comment]
Yep, that is what I am flying at, and with RK5051 props I will have to lower them even more

**Comment by anton-matosov** [comment]
Here is one of the FPV flights https://youtu.be/-SNok411Btc

**Comment by kd0aij** [comment]
Thanks for posting video; that helps a lot for comparison. I'm running 3S, 2300KV motors and 5.4.6 Hyperion bullnose 2-blade props at a field elevation of 1760m. Hard to believe the altitude makes such a big difference; I wonder if something else is different? 

I'd appreciate it if you could post logs using the new logger, with lots of rapid roll/pitch action. (easier to do in stabilize mode without crashing) That would let me run an impulse response analysis for comparison with my racers.

**Comment by anton-matosov** [comment]
You are welcome.
Sure I can do that. How to enable new logger? (I have never used it before)

**Comment by kd0aij** [comment]
Assuming you're using Pixracer, all you have to do is use QGC to set parameter SYS_LOGGER to "new logger".

**Comment by anton-matosov** [comment]
I have just double checked the config and it seems to be wrong.
I will update it now

**Comment by kd0aij** [comment]
Why have you reduced the multirotor mixer roll_scale parameter for the zmr250? Are you using that to compensate for a lower moment of inertia on the roll axis? My approach would be to lower the rate_P gain for the roll axis instead.

**Comment by anton-matosov** [comment]
Are you referring to MC_PITCH_TC and MC_ROLL_TC? these parameters work as simple PID fine tune scale factor (MC_ROLLRATE_P/I/D * 0.2/MC_ROLL_TC, same for pitch). There is no real need to change them, you can simply adjust your PIDs, but I have changed them during original tuning and got current PIDs with TC applied and it was flying really well so I decided to leave them.
These TC params are used by QGroundControl for "roll/pitch sensitivity". I have confused them with RC rates (which are actually "encrypted" in MC_ACRO_R/P/Y_MAX params).

In the next iteration of tuning for 3 blades I am going to remove MC_*_TC from config

**Comment by LorenzMeier** [comment]
Thanks for the good details @anton-matosov, it makes a lot more sense now. It would be great if you could work with @DonLakeFlyer on a tuning UI for multicopters that is more user-friendly (new or different sliders).

**Comment by LorenzMeier** [comment]
I've rebased this, removed the merge commit, removed all noisy (fix ABC) commits and was left with one clean commit. See master.

**Comment by LorenzMeier** [comment]
@anton-matosov Thanks for all the hard work!

File: ROMFS/px4fmu_common/mixers/zmr250.main.mix [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# R: <geometry> <roll scale> <pitch scale> <yaw scale> <deadband>
+R: 4x 7654 10000 10000 0
```
Comment (kd0aij) [inline ROMFS/px4fmu_common/mixers/zmr250.main.mix]: @anton-matosov I'm talking about the roll_scale parameter (7654) on this line

Comment (anton-matosov) [inline ROMFS/px4fmu_common/mixers/zmr250.main.mix]: Ok, never mind my previous post then.
Mixer config is the best way to accommodate custom geometry. ZMR 250 is asymmetric and mixer compensates this.
For ZMR250 mixing for cleanflight/betaflight please refer to https://oscarliang.com/custom-motor-output-mix-quadcopter/

**Comment by anton-matosov** [comment]
I have performed some testing of the acro mode on SK450 frame with 11kv motors, 10x4.5 slow-fly props and 3s battery. And it doesn't seem to have any noticeable oscillations on full throttle punches.
I have not finished PID tuning tho (broke frame), but I have noticed that I can/need get much higher D rate gain compared to racer setup before I get response without overshoot and even much higher gain to get sluggish response or D style twitchiness.
So behavior of PID loop in acro for bigger frame is very different.

**Comment by kd0aij** [comment]
I'd like to see flight logs for both your 250 and 450 quads (in the new
.ulg format) for impulse response analysis.

**Comment by anton-matosov** [comment]
@kd0aij my 450 is flying pixhawk, not pixracer will new logger work there?

**Comment by kd0aij** [comment]
If you're building fmu-v2_default, you have to change the cmake config for that target to build logger instead of sdlog2. Then just set parameter SYS_LOGGER to the new logger and it should start up on boot.

**Comment by anton-matosov** [comment]
@kd0aij here are some flips and rolls logs for my ZMR250 (I hope this is the right file, FlightPlot doesn't plot any of it for some reason)
[ZMR250 flips and rolls logs.zip](https://github.com/PX4/Firmware/files/659270/ZMR250.flips.and.rolls.logs.zip)
Unfortunately I broke my SK450 before I got a chance to grab logs for flips and rolls. So will get them for you once I fix it.

**Comment by kd0aij** [comment]
Thanks, Anton. This log looks fine in FlightPlot; make sure you have the
latest version of PX4/FlightPlot installed.
I'll run the impulse analysis today for comparison with my own 250s. Should
be interesting.

Here's a 450 size frame that would be hard to break:
http://www.aquacopters.com/store/p2/Project_X-Frame_Only.html
but so far I've only crashed mine into water. That saves a lot of props;
I've only managed to break one prop that way.

**Comment by kd0aij** [comment]
@anton-matosov Your tuning looks great! Here's a link to the plots: https://github.com/kd0aij/impulse-response/tree/master/sample_ulogs/ZMR250/flips_250/plots
The rate impulse response is very close to critically damped for both roll and pitch, and the accel tracking has a latency of only 30msec.

**Comment by anton-matosov** [comment]
Thanks. I am very pleased to hear that I did it right 😊

**Comment by anton-matosov** [comment]
The waterproof frame looks interesting. How do you recover your quad from the water? What if it flips over?

**Comment by kd0aij** [comment]
A canoe is a good idea... Even with a receiver antenna taped to the inside bottom of the hull, I've not managed to flip back upright. Seems I lose the RC signal. But I'm sure I've seen videos of copters flipping back upright; would probably work with external whip antennas. A boat (or maybe a Golden Retriever) would still be a good idea though.

**Comment by kd0aij** [comment]
BTW, this is my S250 (3S) tuning: https://github.com/kd0aij/impulse-response/blob/master/sample_ulogs/S250AQ/2016-12-13/plots/rate_impulse.png

A little faster than yours, but a bit under-damped.

**Pull request closed without merging**